rllib/agents/trainer.py

```diff
@@ -764,6 +764,74 @@ def env_creator_from_classpath(env_context):
+        if self.config["exploration_config"].get("type", None) == "RE3":
+            from ray.rllib.evaluation.postprocessing import Postprocessing
+            from ray.rllib.utils.exploration.random_encoder import (
+                update_beta,
+                compute_states_entropy,
+                MovingMeanStd,
+            )
+            from ray.rllib.policy.sample_batch import SampleBatch
+
+            embeds_dim = self.config["exploration_config"].get(
+                "embeds_dim", 128)
+            k_nn = self.config["exploration_config"].get("k_nn", 50)
+            beta = self.config["exploration_config"].get("beta", 0.1)
+            rho = self.config["exploration_config"].get("rho", 0.0001)
+            beta_schedule = self.config["exploration_config"].get(
+                "beta_schedule", "constant")
+
+            class UpdateCallbacks(self.config["callbacks"]):
+                """Update input callbacks method to mutate batch with states entropy rewards."""
+
+                _step = 0
+
+                def __init__(self, *args, **kwargs):
+                    self._rms = MovingMeanStd()
+                    super().__init__(*args, **kwargs)
+
+                def on_learn_on_batch(
+                        self,
+                        *,
+                        policy: Policy,
+                        train_batch: SampleBatch,
+                        result: dict,
+                        **kwargs,
+                ):
+                    states_entropy = compute_states_entropy(
+                        train_batch[SampleBatch.OBS_EMBEDS], embeds_dim, k_nn)
+                    states_entropy = update_beta(
+                        beta_schedule, beta, rho,
+                        UpdateCallbacks._step) * np.reshape(
+                            self._rms(states_entropy),
+                            train_batch[SampleBatch.OBS_EMBEDS].shape[:-1],
+                        )
+                    train_batch[SampleBatch.REWARDS] = (
+                        train_batch[SampleBatch.REWARDS] + states_entropy)
+                    if Postprocessing.ADVANTAGES in train_batch:
+                        train_batch[Postprocessing.ADVANTAGES] = (
+                            train_batch[Postprocessing.ADVANTAGES] +
+                            states_entropy)
+                        train_batch[Postprocessing.VALUE_TARGETS] = (
+                            train_batch[Postprocessing.VALUE_TARGETS] +
+                            states_entropy)
+                    super().on_learn_on_batch(
+                        policy=policy,
+                        train_batch=train_batch,
+                        result=result,
+                        **kwargs)
+
+                def on_train_result(self, *, trainer, result: dict,
+                                    **kwargs) -> None:
+                    # Keep track of the training iteration for beta decay.
+                    UpdateCallbacks._step = result["training_iteration"]
```

Comment:
I think you can just use policy.global_timestep and there is no need to track this yourself?
another way is to Add a MixIn for this Beta schedule, much like LearningRateSchedule and EntropyCoeffSchedule. you can then just use the auto-decayed beta value here.

Reply:
Here, we want to decay based on `training_iteration`. It seems `policy.global_timestep` depends on number of batch processed and not equal to `training_iteration`, just looking at source, not 100% sure.